Updated docs

tarkatronic opened the pull request:
This is the beginning of a rework of the documentation for tartufo. It has had many issues for some time now, so I am taking this opportunity to incorporate user feedback and improve the overall structure and content of the documentation.

I am starting this as a draft, to give people a sense of where I am headed. The first swing I am taking is at making the documentation more complete. Next I will work on improving the overall structure, so that we don't have a confusing mess of "Usage" vs "Features" vs "Configuration" (that barely even talks about configuration). And then finally, adding an upgrade guide from v2 -> v3.

tarkatronic committed:
Fix up markdownlint warnings

tarkatronic committed:
Fix up docs build warnings

tarkatronic committed:
Update ToC

tarkatronic committed:
Ensure all the util docs are complete

tarkatronic committed:
Use sphinx_autodoc_typehints and update intersphinx mapping

tarkatronic committed:
Update all sphinx-related tooling to the latest

tarkatronic committed:
Fix up spelling problems

tarkatronic committed:
Improve the documentation of the types module

tarkatronic committed:
Add missing Union import

tarkatronic committed:
Improve the scanner docstrings

tarkatronic committed:
Finish up the types docstrings

tarkatronic committed:
Fix docstrings for util

tarkatronic committed:
Shift more configuration options from features -> config

sushantmimani committed:
Update config docs

rbailey-godaddy commented on CHANGELOG.md:
```diff
@@ -126,7 +126,7 @@ Features:
   Users who have the old style configuration will now receive a
   `DeprecationWarning` stating that the old behavior will go away with v3.0.
-* Fixed up warning handling so that we can display `DeprecationWarning`s to users
+* Fixed up warning handling so that we can display `DeprecationWarnings` to users
```
This is probably more correct the original way, as the `s` is intended to be plural rather than part of the type.

rbailey-godaddy commented on CHANGELOG.md:
```diff
@@ -89,7 +89,7 @@ Misc:
 
 v2.10.0 - 3 November 2021
-------------
+-------------------------
```
Do we need or want to merge the v2.10.1 notes here too?

rbailey-godaddy commented on docs/source/configuration.rst:
```diff
@@ -105,5 +214,54 @@ Option 2: Separating each rule out into its own table.
+path-pattern No       Regular expression           A pattern to specify to what files the exclusion will apply
+reason       No       String                       A plaintext reason the exclusion has been added
+match-type   No       String ("match", or "scope") Whether to perform a `search or match`_ regex operation
+scope        No       String ("word", or "line")   Whether to match against the current word or full line of text
```
```suggestion
match-type   No       String ("match" or "scope")  Whether to perform a `search or match`_ regex operation
scope        No       String ("word" or "line")    Whether to match against the current word or full line of text
```

rbailey-godaddy reviewed (changes requested):
Nits. Also, asking again more concretely -- should I add commits for the v2-to-v3 upgrade documentation to this PR, or start a new one after this is merged?

rscottbailey committed:
Add "Upgrading" documentation

This basically expands the CHANGELOG content from 3.x into a few sets of

* What will I see different?
* What do I do about it?

topics that should get users over the 2-to-3 hump even if they read
nothing else.

sushantmimani committed:
Update docs/source/configuration.rst

Co-authored-by: Scott Bailey <72747501+rbailey-godaddy@users.noreply.github.com>

sushantmimani committed:
Address review comments

sushantmimani committed:
Hide unsupported options

rscottbailey committed:
Add "oops not there yet" section

rscottbailey committed:
Merge branch 'updated-docs' of ghc:godaddy/tartufo into updated-docs

rbailey-godaddy reviewed (approved):
Of course I think everything I wrote is fabulous. ;-) Everything else LGTM too...

sushantmimani commented:
> Of course I think everything I wrote is fabulous. ;-) Everything else LGTM too...

Ditto!

sushantmimani reviewed (approved):
LGTM 🦅

rbailey-godaddy commented:
Fixes #279
